aten/src/ATen/native/sparse/SparseTensorMath.cpp

```diff
@@ -205,83 +205,30 @@ SparseTensor& add_out_sparse_cpu(SparseTensor& r, const SparseTensor& t, const S
+  // We deliberately choose to simply concat the indices and values tensors
+  // rather than merging them. This makes `add_out_sparse_cpu` work on sparse tensors
+  // with non-contiguous `values`, at the cost of having a non-coalesced result.
+  // This trade-off is preferable for the common use-case of gradient accumulation.
 
   AT_DISPATCH_ALL_TYPES(
-      t_values.scalar_type(), "cadd_sparse", [&] {
-        scalar_t* t_values_ptr = t_values.data<scalar_t>();
-        scalar_t* s_values_ptr = s_values.data<scalar_t>();
-        scalar_t* r_values_ptr = r_values.data<scalar_t>();
-        scalar_t cast_value = value.to<scalar_t>();
-        while (t_i < t_nnz || s_i < s_nnz) {
-          if (t_i >= t_nnz) {
-            cmp = -1;
-          } else if (s_i >= s_nnz) {
-            cmp = 1;
-          } else {
-            cmp = 0;
-            for (d = 0; d < sparse_dim; d++) {
-              if (t_indices_accessor[d][t_i] < src_indices_accessor[d][s_i]) {
-                cmp = 1;
-                break;
-              }
-              if (t_indices_accessor[d][t_i] > src_indices_accessor[d][s_i]) {
-                cmp = -1;
-                break;
-              }
-            }
-          }
-          if (cmp >= 0) {
-            for (d = 0; d < sparse_dim; d++) {
-              r_indices_accessor[d][r_i] = t_indices_accessor[d][t_i];
-            }
-            if (t_values.numel() > 0) {  // We add all elements from t_values to r_values only if t_values is not an empty tensor
-              THBlas_axpy<scalar_t>(blockSize, 1,
-                t_values_ptr + t_i * blockSize, 1,
-                r_values_ptr + r_i * blockSize, 1);
-            }
-            t_i++;
-          }
-          if (cmp <= 0) {
-            for (d = 0; d < sparse_dim; d++) {
-              r_indices_accessor[d][r_i] = src_indices_accessor[d][s_i];
-            }
-            if (s_values.numel() > 0) {  // We add all elements from s_values to r_values only if s_values is not an empty tensor
-              THBlas_axpy<scalar_t>(blockSize, cast_value,
-                s_values_ptr + s_i * blockSize, 1,
-                r_values_ptr + r_i * blockSize, 1);
-            }
-            s_i++;
-          }
-          r_i++;
+    s_values.scalar_type(), "add_out_sparse_cuda", [&] {
+        if (value.to<scalar_t>() != static_cast<scalar_t>(1)) {
+          s_values = s_values.mul(value);
         }
-      }
-  );
+      });
 
-  get_sparse_impl(r)->set_nnz_and_narrow(r_i);
-  // TODO: I think it may be possible to track inside the loop and
-  // detect when we are uncoalesced (e.g., by observing that an
-  // index goes backwards) which may be more precise than using the
-  // coalesced flag here.  But this is easy.
-  return r._coalesced_(t_coalesced && s_coalesced);
+  LongTensor r_indices = at::cat({t_indices, s_indices}, 1);
+  Tensor r_values = at::cat({t_values, s_values}, 0);
+  alias_into_sparse(r, r_indices, r_values);
```

Comment:
If you cat'ed, don't you have to specify the output is not coalesced

Reply:
IMO we should make this a parameter on `alias_into_sparse` so people have to consider it.

Reply:
`alias_into_sparse(...)` calls `set_indices_and_values_unsafe(...)` internally which always sets `coalesced_ = false`, and we expect users to call `sparse_tensor._coalesced_(...)` afterwards if they want to change the coalesce-ness of the sparse tensor. For example: https://github.com/pytorch/pytorch/blob/1c671c56c16b203ececd1ee6f70a0f6fcb9265a5/aten/src/ATen/native/sparse/SparseTensor.cpp#L457-L458

To simplify this API, we can add an `is_coalesced` parameter on `alias_into_sparse`, possibly in a separate PR.